receiver/kafkareceiver/kafka_receiver.go

```diff
@@ -38,310 +40,279 @@ const (
-		// recreated to get the new claims
-		if err := c.consumerGroup.Consume(ctx, c.topics, handler); err != nil {
-			c.settings.Logger.Error("Error from consumer", zap.Error(err))
-		}
-		// check if context was cancelled, signaling that the consumer should stop
-		if ctx.Err() != nil {
-			c.settings.Logger.Info("Consumer stopped", zap.Error(ctx.Err()))
-			return
-		}
-	}
+func (h *tracesHandler) endObsReport(ctx context.Context, n int, err error) {
+	h.obsrecv.EndTracesOp(ctx, h.encoding, n, err)
 }
 
-func (c *kafkaMetricsConsumer) Shutdown(context.Context) error {
-	if c.cancelConsumeLoop == nil {
-		return nil
-	}
-	c.cancelConsumeLoop()
-	c.consumeLoopWG.Wait()
-	if c.consumerGroup == nil {
-		return nil
+func (h *tracesHandler) getResources(data ptrace.Traces) iter.Seq[pcommon.Resource] {
+	return func(yield func(pcommon.Resource) bool) {
+		for _, rm := range data.ResourceSpans().All() {
+			if !yield(rm.Resource()) {
+				return
+			}
+		}
 	}
-	return c.consumerGroup.Close()
 }
 
-func newLogsReceiver(config *Config, set receiver.Settings, nextConsumer consumer.Logs) (*kafkaLogsConsumer, error) {
+func newKafkaConsumer(
+	config *Config,
+	set receiver.Settings,
+	topics []string,
+	newConsumeMessageFunc func(*consumerGroupHandler, component.Host) (consumeMessageFunc, error),
+) (*kafkaConsumer, error) {
 	telemetryBuilder, err := metadata.NewTelemetryBuilder(set.TelemetrySettings)
 	if err != nil {
 		return nil, err
 	}
 
-	return &kafkaLogsConsumer{
-		config:            config,
-		topics:            []string{config.Logs.Topic},
-		nextConsumer:      nextConsumer,
-		consumeLoopWG:     &sync.WaitGroup{},
-		settings:          set,
-		autocommitEnabled: config.AutoCommit.Enable,
-		messageMarking:    config.MessageMarking,
-		headerExtraction:  config.HeaderExtraction.ExtractHeaders,
-		headers:           config.HeaderExtraction.Headers,
-		telemetryBuilder:  telemetryBuilder,
-		minFetchSize:      config.MinFetchSize,
-		defaultFetchSize:  config.DefaultFetchSize,
-		maxFetchSize:      config.MaxFetchSize,
+	return &kafkaConsumer{
+		config:                config,
+		topics:                topics,
+		newConsumeMessageFunc: newConsumeMessageFunc,
+		settings:              set,
+		telemetryBuilder:      telemetryBuilder,
 	}, nil
 }
 
-func (c *kafkaLogsConsumer) Start(_ context.Context, host component.Host) error {
-	ctx, cancel := context.WithCancel(context.Background())
-	c.cancelConsumeLoop = cancel
+// kafkaConsumer consumes messages from a set of Kafka topics,
+// decodes telemetry data using a given unmarshaler, and passes
+// them to a consumer.
+type kafkaConsumer struct {
+	config                *Config
+	topics                []string
+	settings              receiver.Settings
+	telemetryBuilder      *metadata.TelemetryBuilder
+	newConsumeMessageFunc func(*consumerGroupHandler, component.Host) (consumeMessageFunc, error)
+
+	mu                sync.Mutex
+	started           bool
+	shutdown          bool
```

Comment:
Could these be replaced with atomic bools to save needing the mutex?

Reply:
It's technically possible to get rid of the mutex, but the mutex also guards access to consumerGroup and consumeLoopClosed, so to do it safely you would need additional atomics and probably some busy-waiting.

Reply:
Resolving, but happy to discuss suggestions